File: server/src/ai/pdfProcessor.js
import fs from 'fs';
import path from 'path';
import crypto from 'crypto';

const UPLOADS_DIR = path.join(process.cwd(), 'server', 'uploads');

export function ensureUploadsDir() {
  if (!fs.existsSync(UPLOADS_DIR)) {
    fs.mkdirSync(UPLOADS_DIR, { recursive: true });
  }
  return UPLOADS_DIR;
}

export function getFilePath(filename) {
  return path.join(UPLOADS_DIR, filename);
}

export async function extractTextFromPDF(pdfPath) {
  try {
    const { extractText } = await import('unpdf');
    const fileBuffer = fs.readFileSync(pdfPath);
    const pdf = await extractText(fileBuffer);
    return pdf.text || '';
  } catch (err) {
    console.error('PDF extraction error:', err.message);
    return null;
  }
}

export function detectChapters(text) {
  const chapterPatterns = [
    /(?:chapter|unit|module|lesson)\s*[:\s]*(?:[0-9]+|[IVXLCDM]+)\s*[:\s.-]*([^\n]+)/gi,
    /^([A-Z][A-Z\s]{2,50})$/gm,
    /^(\d+\.\d+\s+[A-Z][^\n]{2,100})$/gm
  ];

  const chapters = [];
  const lines = text.split('\n');

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (!line) continue;

    for (const pattern of chapterPatterns) {
      pattern.lastIndex = 0;
      const match = pattern.exec(line);
      if (match) {
        const name = (match[1] || match[0]).trim().substring(0, 200);
        if (name.length > 3 && !chapters.find(c => c.name === name)) {
          chapters.push({ name, lineNumber: i + 1 });
        }
        break;
      }
    }
  }

  return chapters;
}

export function chunkText(text, chunkSize = 1000, overlap = 200, chapters = []) {
  const paragraphs = text.split(/\n\s*\n/);
  const chunks = [];
  let currentChunk = '';
  let currentChapter = 'Introduction';
  let currentPage = 1;
  let chunkIndex = 0;

  const pagePattern = /page\s*(\d+)|\[page\s*(\d+)\]|(\d+)\s*\[page\]/gi;

  for (let para of paragraphs) {
    para = para.trim();
    if (!para) continue;

    for (const ch of chapters) {
      if (para.toLowerCase().includes(ch.name.toLowerCase())) {
        currentChapter = ch.name;
      }
    }

    const pageMatch = pagePattern.exec(para);
    if (pageMatch) {
      currentPage = parseInt(pageMatch[1] || pageMatch[2] || pageMatch[3]) || currentPage;
    }

    if (currentChunk.length + para.length > chunkSize && currentChunk.length > 0) {
      chunks.push({
        index: chunkIndex,
        text: currentChunk.trim(),
        chapter: currentChapter,
        page: currentPage
      });
      chunkIndex++;
      const words = currentChunk.split(' ');
      const overlapWords = words.slice(-Math.floor(overlap / 5)).join(' ');
      currentChunk = overlapWords + '\n\n' + para;
    } else {
      currentChunk += (currentChunk ? '\n\n' : '') + para;
    }
  }

  if (currentChunk.trim()) {
    chunks.push({
      index: chunkIndex,
      text: currentChunk.trim(),
      chapter: currentChapter,
      page: currentPage
    });
  }

  return chunks;
}

export function generateChunkId(textbookId, chunkIndex) {
  return `${textbookId}_${chunkIndex}`;
}

export async function saveUploadedFile(file) {
  ensureUploadsDir();
  const ext = path.extname(file.originalname || '.pdf');
  const filename = `${Date.now()}_${crypto.randomBytes(4).toString('hex')}${ext}`;
  const filepath = path.join(UPLOADS_DIR, filename);

  if (file.buffer) {
    fs.writeFileSync(filepath, file.buffer);
  } else if (file.path) {
    fs.copyFileSync(file.path, filepath);
  }

  return {
    filename,
    filepath,
    extension: ext,
    size: file.size || (file.buffer ? file.buffer.length : 0)
  };
}

export default { extractTextFromPDF, detectChapters, chunkText, generateChunkId, saveUploadedFile, ensureUploadsDir };

File: backend/app/core/pdf_processor.py
import os
import fitz
import re
from typing import Optional


class PDFProcessor:
    def __init__(self, chunk_size: int = 500, overlap: int = 50):
        self.chunk_size = chunk_size
        self.overlap = overlap

    def extract_text(self, pdf_path: str) -> Optional[str]:
        try:
            doc = fitz.open(pdf_path)
            text = ""
            for page in doc:
                text += f"\n\n[Page {page.number + 1}]\n"
                text += page.get_text()
            doc.close()
            return text
        except Exception as e:
            print(f"PDF extraction error: {e}")
            return None

    def chunk_text(self, text: str, chapter_metadata: dict = None) -> list:
        paragraphs = re.split(r'\n\s*\n', text)
        chunks = []
        current_chunk = ""
        current_page = 1
        current_chapter = chapter_metadata.get("name", "Introduction") if chapter_metadata else "Introduction"
        chunk_index = 0
        page_match_pattern = re.compile(r'\[Page\s*(\d+)\]')

        for para in paragraphs:
            para = para.strip()
            if not para:
                continue

            page_match = page_match_pattern.search(para)
            if page_match:
                current_page = int(page_match.group(1))

            if len(current_chunk) + len(para) > self.chunk_size and current_chunk:
                chunks.append({
                    "text": current_chunk.strip(),
                    "page": current_page,
                    "chapter": current_chapter,
                    "chunk_index": chunk_index
                })
                chunk_index += 1
                words = current_chunk.split()
                overlap_words = " ".join(words[-max(1, self.overlap // 5):])
                current_chunk = overlap_words + "\n\n" + para if overlap_words else para
            else:
                current_chunk += ("\n\n" + para) if current_chunk else para

        if current_chunk.strip():
            chunks.append({
                "text": current_chunk.strip(),
                "page": current_page,
                "chapter": current_chapter,
                "chunk_index": chunk_index
            })

        return chunks

    def detect_chapters(self, text: str) -> list:
        patterns = [
            r'(?:Chapter|Unit|Module|Lesson)\s*[:\s]*(?:[0-9]+|[IVXLCDM]+)\s*[:\s.-]*([^\n]+)',
            r'^([A-Z][A-Z\s]{2,50})$',
            r'^(\d+\.\d+\s+[A-Z][^\n]{2,100})$'
        ]

        chapters = []
        lines = text.split("\n")

        for i, line in enumerate(lines):
            line = line.strip()
            if not line:
                continue
            for pattern in patterns:
                match = re.search(pattern, line, re.IGNORECASE | re.MULTILINE)
                if match:
                    name = (match.group(1) or match.group(0)).strip()[:200]
                    if len(name) > 3 and name not in [c["name"] for c in chapters]:
                        chapters.append({"name": name, "line_number": i + 1})
                    break

        return chapters

    def process_ncert_pdf(self, pdf_path: str, subject: str, chapter_metadata: dict) -> list:
        text = self.extract_text(pdf_path)
        if not text:
            return []

        chunks = self.chunk_text(text, chapter_metadata)

        for chunk in chunks:
            chunk["subject"] = subject
            chunk["chapter"] = chapter_metadata.get("name", chunk["chapter"])

        return chunks
